### ISSUE TITLE
Infinite type not detected

### ISSUE THREAD

**Issue by evancz** [elm-lang/elm-compiler#707]
Continuation of #540, stack overflow for the following code because of an infinite type:

```elm
import Either (..)

data Automaton a b = Step (a -> (Automaton a b, b))

(+++) : Automaton a b -> Automaton c d -> Automaton (Either a c) (Either b d)
(Step l) +++ (Step r) = Step <| \o ->
    case o of
        Left a -> 
            let (l1, c) = l a
            in (l1, Left c)
            
        Right b -> 
            let (r1, d) = r b
            in (r1, Right d)
```

Fixed code has corrected the return type of both branches to keep things going properly:

```haskell
import Either (..)

data Automaton a b = Step (a -> (Automaton a b, b))

(+++) : Automaton a b -> Automaton c d -> Automaton (Either a c) (Either b d)
(Step l) +++ (Step r) = Step <| \o ->
    case o of
        Left a -> 
            let (l1, c) = l a
            in (l1 +++ Step r, Left c)
            
        Right b -> 
            let (r1, d) = r b
            in (Step l +++ r1, Right d)
```

Probably best as an example of "when things go wrong".

**Comment by evancz** [comment]
Closing in favor of #832
